### PR TITLE
fix metrics to ensure cpu vs gpu

### PR DESCRIPTION
**Pull request by joelgrus**
I believe this should fix 

https://github.com/allenai/allennlp/issues/166

I didn't change the current example because it's only copying one axis (?) and it didn't seem like there was an easy way to generalize that

### CODE REVIEW

**Comment by matt-gardner** [comment]
Did you try running this?  Doesn't look to me like it should work.  The issue is that you're creating the tensor (with `torch.ones`), then copying it on _the same device_.  You need to start from the _mask_, copy it, then fill it with ones.

**Comment by matt-gardner** [comment]
A useful function would be something like `ones_like`, that creates a tensor of ones of the same shape and on the same device as a given tensor.  Then you call `ones_like(mask)`, and you're good.

**Comment by joelgrus** [comment]
so thinking about this more, I don't know if we even need some of these variables.  For example, in CategoricalAccuracy, couldn't we just do

```
        if mask is not None:
            correct *= mask.unsqueeze(-1)
            self.total_count += mask.sum()
        else:
            self.total_count += gold_labels.numpy().size
```

and then not need the `count` variable at all?

and possibly the f1 ones don't need any changing at all? it seems like they work as is

**Comment by matt-gardner** [comment]
Yeah, I think you're right about being able to just get rid of the intermediate variable.  I'd use [`gold_labels.numel()`](http://pytorch.org/docs/master/torch.html#torch.numel) instead of `.numpy().size`.  And I wouldn't be confident that the f1 metric works as it is - Mark is using a different f1 metric for the SRL model, so nothing is currently testing that metric on a GPU, and it looks like it's similarly broken.

**Comment by joelgrus** [comment]
ok, this version should be correct, I tested the F1 on a CPU and it works

File: allennlp/nn/util.py [hunk 1]
```diff
@@ -392,3 +392,10 @@ def inner_device_mapping(storage: torch.Storage, location) -> torch.Storage:  #
         else:
             return storage
     return inner_device_mapping
+
+def ones_like(tensor: torch.Tensor) -> torch.Tensor:
+    """
+    Use clone() + copy_() to make sure that a tensor ends up on the right
+    device at runtime.
+    """
+    return tensor.clone().copy_(torch.ones(tensor.shape))
```
Comment (matt-gardner) [inline allennlp/nn/util.py]: I think it'd be more efficient to just use `.fill_(1)` instead of the `.copy()` here.
